src/main/java/com/google/code/or/common/util/MySQLUtils.java

```diff
@@ -79,6 +79,13 @@ public static int toYear(int value) {
         return new java.sql.Time(millis + (nanosForFractionalValue(fraction, width) / 1000000));
 	}
 
+	public static java.sql.Timestamp toTimestamp2FromRawTime(int value, int fraction, int width) {
+		final java.sql.Time time = toTime2(value, fraction, width);
+		final java.sql.Timestamp t = new java.sql.Timestamp(time.getTime());

```

Comment:
no, ok, I see where you're going now;  there's a lot of deep juju about the binlog format encoded in the `toTime2` functions.  I was getting tripped up in the weird naming, mostly.

I think the correct thing to do here is to do a bit of minor refactoring:

```
   public Timestamp time2ToTimestamp(int value, int fraction, int width)
       Timestamp t  = new Timestamp(getMillisFromTime2(value));
       t.setNanos(...)
       return t;

   public long getMillisFromTime2(value) { 
	final int h = (value >> 12) & 0x3FF;
        ...
        return c.getTimeInMillis();
   }

   public Timestamp datetime2ToTimestamp(int value, int fraction, int width)
   public long getMillisFromDatetime2(value) { 
   public Timestamp timestamp2ToTimestamp(int value, int fraction, int width)
   public long getMillisFromTimestamp2(value) { 

```

So you end up with 3 functions to take `value` and turn it into milliseconds, 1 function to add nanos, and 3 functions to convert the rest to a `Timestamp`.

reasonable enough?